spec/common/testapp_module.js

```diff
@@ -0,0 +1,79 @@
+/* global console, jasmine, define, beforeAll, afterAll, global, __ABLY__, Ably */
+"use strict";
+
+/* Shared test helper used within Jasmine tests */
+
+define(['testapp'], function(testAppModule) {
+  var isBrowser = (typeof(window) === 'object');
+
+  function globalObject() {
+    if (isBrowser) {
+      return window;
+    } else {
+      return global;
+    }
+  }
+
+  function updateTestApp(newTestApp) {
+    globalObject().AblyTestApp = newTestApp;
+  }
+
+  function getTestApp() {
+    return globalObject().AblyTestApp;
+  }
+
+  function incrementBeforeAllCounter(quantity) {
+    if (!globalObject().AblyBeforeAllCounter) { globalObject().AblyBeforeAllCounter = 0; }
+    globalObject().AblyBeforeAllCounter += quantity;
+    return globalObject().AblyBeforeAllCounter;
+  }
+
+  function getBeforeAllCounter() {
+    return globalObject().AblyBeforeAllCounter;
+  }
+
+  function setup(done) {
+    incrementBeforeAllCounter(1);
+    if (getTestApp()) { return done(); }
+
+    testAppModule.setup(function(err, newTestApp) {
+      if (err) {
+        throw "Could not set up Test App: " + JSON.stringify(err);
+      } else {
+        updateTestApp(newTestApp);
+        console.log("Test App " + getTestApp().appId + " has been set up");
+        done();
+      }
+    });
+  }
+
+  function tearDown(done) {
+    if (incrementBeforeAllCounter(-1) <= 0) { return done(); } // tearDown only if last afterAll block to run
```

Comment:
I know, I didn't want to introduce a new method and couldn't think of a better name at 4 in the morning!